### PR TITLE
OCLOMRS-472: Implement the Retire/Un-Retire button

### PR DESCRIPTION
**Pull request by brucemakallan**
# JIRA TICKET NAME:
[Implement the Retire/Un-Retire button](https://issues.openmrs.org/browse/OCLOMRS-472)

# Summary:
Ensure that one can Retire custom Concepts as well as Unretire them

### CODE REVIEW

**Comment by brucemakallan** [comment]
This PR is currently a Work In Progress

File: package.json [hunk 1]
```diff
@@ -30,7 +30,7 @@
     "uuid": "^3.3.2"
   },
   "scripts": {
-    "start": "chmod +x ./init.sh && ./init.sh && cp env-config.js ./public/ && npm run build:css && react-scripts start",
```
Comment (dkayiwa) [inline package.json]: How is this related to the ticket?

Comment (brucemakallan) [reply package.json]: I have created a separate ticket for it. I'll remove the change from this one

File: src/tests/Dashboard/action/dictionaryAction.test.js [hunk 2]
```diff
@@ -381,6 +382,35 @@ describe('Test suite for dictionary actions', () => {
...
+  it('should retire/unretire a concept', () => {
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith({
+        status: 200,
+        response: sampleConcept,
+      });
+    });
+    const store = mockStore({ payload: {} });
+    return store.dispatch(retireConcept(sampleConcept.url, { retired: false })).then(() => {
```
Comment (dkayiwa) [inline src/tests/Dashboard/action/dictionaryAction.test.js]: How is this testing for should unretire a concept?

Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: I have updated the `retireConcept()` action to return the retired concept and I am now testing directly for the retired flag. `expect(result.retired).toEqual(true);` when retiring, and `expect(result.retired).toEqual(false);` when un-retiring.

File: src/tests/Dashboard/action/dictionaryAction.test.js [hunk 3]
```diff
@@ -381,6 +382,49 @@ describe('Test suite for dictionary actions', () => {
       expect(store.getActions()).toEqual(expectedActions);
     });
   });
+
+  it('should retire a concept', () => {
```
Comment (dkayiwa) [inline src/tests/Dashboard/action/dictionaryAction.test.js]: When should retire a concept happen?

File: src/tests/Dashboard/action/dictionaryAction.test.js [hunk 4]
```diff
@@ -381,6 +382,49 @@ describe('Test suite for dictionary actions', () => {
...
+        response: sampleRetiredConcept,
+      });
+    });
+    const store = mockStore({ payload: {} });
+    return store.dispatch(retireConcept(sampleConcept.url, { retired: true })).then((result) => {
+      expect(result.retired).toEqual(true);
+    });
+  });
+
+  it('should unretire a concept', () => {
```
Comment (dkayiwa) [inline src/tests/Dashboard/action/dictionaryAction.test.js]: When should unretire a concept happen?

Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: This action is triggered when the `unretire` button is clicked

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 3 above)
Comment (dkayiwa) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: Will you keep around to keep answering the maintainers of this code whenever they ask this?

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 4 above)
Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: I have updated the message accordingly.

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 3 above)
Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: I have updated the message accordingly

File: src/tests/Dashboard/action/dictionaryAction.test.js [hunk 5]
```diff
@@ -381,6 +382,49 @@ describe('Test suite for dictionary actions', () => {
...
+
+  it('should retire a concept when the retire button is clicked', () => {
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith({
+        status: 200,
+        response: sampleRetiredConcept,
+      });
+    });
+    const store = mockStore({ payload: {} });
```
Comment (dkayiwa) [inline src/tests/Dashboard/action/dictionaryAction.test.js]: Where are is the simulation of clicking the retire button?

File: src/tests/Dashboard/action/dictionaryAction.test.js [hunk 6]
```diff
@@ -381,6 +382,49 @@ describe('Test suite for dictionary actions', () => {
...
+
+  it('should unretire a concept when the unretire button is clicked', () => {
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith({
+        status: 200,
+        response: sampleConcept,
+      });
+    });
+    const store = mockStore({ payload: {} });
```
Comment (dkayiwa) [inline src/tests/Dashboard/action/dictionaryAction.test.js]: Same as above

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 5 above)
Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: `DictionaryConcepts.test.jsx` lines 528 - 621

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 6 above)
Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: `DictionaryConcepts.test.jsx` lines 528 - 621

File: src/redux/actions/dictionaries/dictionaryActionCreators.js [hunk 7]
```diff
@@ -18,6 +18,7 @@ import {
 import { filterPayload } from '../../reducers/util';
 import { addDictionaryReference } from '../bulkConcepts';
 import api from '../../api';
+import instance from '../../../config/axiosConfig';
```
Comment (dkayiwa) [inline src/redux/actions/dictionaries/dictionaryActionCreators.js]: What is the meaning of instance?

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 5 above)
Comment (dkayiwa) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: I mean in this very test?

Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: This is the Action that is triggered when the button is clicked. The test is only for the Action and the button click test is in the lines I mentioned. Therefore the simulation for the button click is not (and should not be) in this test.

File: src/redux/actions/dictionaries/dictionaryActionCreators.js (hunk 7 above)
Comment (brucemakallan) [reply src/redux/actions/dictionaries/dictionaryActionCreators.js]: I have used a more appropriate name (`axiosInstance`)

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 5 above)
Comment (dkayiwa) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: Then i find this confusing "should retire a concept when the retire button is clicked"

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 6 above)
Comment (dkayiwa) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: Same as above

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 5 above)
Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: I understand the concern. Should I change it to `"should retire a concept when the retireConcept method is triggered"`?

Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: I have changed the title to `should retire a concept when the retireConcept action is triggered with the true argument`

File: src/tests/Dashboard/action/dictionaryAction.test.js (hunk 6 above)
Comment (brucemakallan) [reply src/tests/Dashboard/action/dictionaryAction.test.js]: I have changed the title to `should unretire a concept when the retireConcept action is triggered with the false argument`
